refactor: standardize trip filtering; add an invariant model integration test

dmuldrew opened the pull request:
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose

Incorporates changes from @danlivengood to match “immediate” and “smart charging” daily charging output results. Replaces existing integration test with a dimensionless invariant calculation that should equal 1 for the model.

### What the code is doing

Standardizes the trip filtering between immediate and smart charging implementations.

Also adds the following automated test:
``` python
abs(((output_load_sum_array.sum() * charging_efficiency) / (daily_vmt_array.sum() * kwhmi) ) - 1) < 0.03
```
which checks that the daily load output over the trip window (adjusted by charging efficiency) equals the vehicles miles travel times a kWh per mile factor.

### Testing

Manual and automated testing

### Where to look

Files in the transportation electrification folder

### Time estimate

~1hr

dmuldrew committed:
refactor: update daily demand output scaling

dmuldrew committed:
refactor: add cyclic trip dataframe filter

dmuldrew committed:
fix: add missing loop iterator; add missing trip filters

dmuldrew committed:
test: add invariant tests

dmuldrew committed:
refactor: adapt profile interface for tuple output

fix: separate trip filter ops to pass integration test

rouille commented:
You might want to change the target branch to `transportation_electrification`

rouille commented on prereise/gather/demanddata/transportation_electrification/data_helper.py:
```diff
@@ -309,37 +309,20 @@ def get_total_daily_vmt(data: pd.DataFrame, input_day, daily_values):
     :return: (*np.array*) -- an array where each element is the daily VMT and total
         vehicles for that day.
     """
```
outdated docstring: `daily_values` --> `veh_type`, name, type and description need to be updated

rouille replied:
Also, a `:raises ValueError:` entry needs to be added

rouille commented on prereise/gather/demanddata/transportation_electrification/immediate.py:
```diff
@@ -207,6 +250,8 @@ def immediate_charging(
     for i, weekday_flag in enumerate(input_day):
         daily_profile = daily_resampled_profiles[flag_translation[weekday_flag]]
+        # print(f"day: {i}")
+        # print(f"daily sum: {np.sum(daily_profile)}")
```
Do we want to keep these comments?

rouille commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -68,19 +74,26 @@ def smart_charging(
     # load NHTS data from function
     if veh_type.lower() == "ldv":
-        newdata = data_helper.remove_ldt(data_helper.load_data(census_region, filepath))
+        newdata1 = data_helper.remove_ldt(
```
Why use `newdata1` as variable name?

danlivengood replied:
That was something I did in case the original dataframe shouldn't be updated via the filtering process we added.
Now that we have automated tests, **and** as long as my concern is outdated, then I'm on board with only using `newdata` and not creating `newdata1`

rouille commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -156,6 +254,34 @@ def smart_charging(
+        #             "energy limit",
+        #             "rates",
+        #         ]
+        #     )
```
Do we need to keep these comments?

danlivengood replied:
Yes.  Ultimately, we want to uncomment these to include as a debugging option for the user / developer.  Unfortunately, at the moment, we ran into an error with `pd.concat` on line 449 that does not impact the profile generation but does impact this bonus feature.

rouille commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -298,20 +438,99 @@ def smart_charging(
+                    # print(individual.index)
+                    # missed_vehicles = pd.concat(
+                    #     [missed_vehicles, individual], ignore_index=True
+                    # )
```
Do we need to keep these comments

danlivengood replied:
Same reply as above

rouille commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -298,20 +438,99 @@ def smart_charging(
+            #             # "trip_miles","Vehicle miles traveled","why_from",
+            #         ]
+            #     ]
+            # )
```
Do we need to keep these comments?

danlivengood replied:
Same reply as above

dmuldrew committed:
doc: update existing function docs strings

dmuldrew committed:
refactor: simplify trip data names

fix: missed_vehicles concatenation

refactor: refine filtering; unneeded print statements

fix: add nd_len back where needed

refactor: parameter specification, indexing refinement

chore: linter complaints resolved

BainanXia commented on prereise/gather/demanddata/transportation_electrification/data_helper.py:
```diff
@@ -297,49 +297,32 @@ def generate_daily_weighting(year, area_type="urban"):
-    return daily_vmt_total
-
-
-def get_total_hdv_daily_vmt(data: pd.DataFrame, veh_range):
```
Just to confirm, we move the vehicle range logic into the charging function for immediate charging which mirrors what we do for smart charging, right?

danlivengood replied:
Correct.  When this function is called, the vehicle range filter has already been applied and thus is no longer used in this function.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/immediate.py:
```diff
@@ -109,6 +110,7 @@ def immediate_charging(
     veh_type,
     filepath,
     trip_strategy=1,
+    input_day=None,
```
We will need to update the doc strings for this parameter below.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/immediate.py:
```diff
@@ -133,10 +135,36 @@ def immediate_charging(
+    del trips
+    trips = filtered_census_data
+
+    print(f"updated data length:{nd_len}")
```
It seems `nd_len` is not changed since defined, or I'm missing something.

danlivengood replied:
Good catch.  This was part of our debugging process more than anything.  @dmuldrew, do we need this print statement anymore?
If we do, like Bainan noticed, we should recalculate `nd_len` to capture any changes from the cyclical trips filtering

BainanXia commented on prereise/gather/demanddata/transportation_electrification/immediate.py:
```diff
@@ -133,10 +135,36 @@ def immediate_charging(
+        individual = trips.iloc[i : i + total_trips].copy()
+        if (
+            individual["why_from"].head(1).values[0]
+            == individual["dwell_location"].tail(1).values[0]
```
Maybe we can simply do `individual["why_from"].iloc[0] == individual["dwell_location"].iloc[-1]`

BainanXia commented on prereise/gather/demanddata/transportation_electrification/immediate.py:
```diff
@@ -133,10 +135,36 @@ def immediate_charging(
+                [filtered_census_data, individual], ignore_index=True
+            )
+        i += total_trips
+    del trips
```
Not sure this is necessary unless we mean to release some memory immediately. As for as I know removing the reference won't reclaim the memory immediately unless there is zero reference on a certain chunk of memory which can be handled well by Python in the background.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/immediate.py:
```diff
@@ -182,14 +210,29 @@ def immediate_charging(
+        trips["charging_allowed"] * trips["dwell_time"] * power * charging_efficiency
+    )
+
+    grouped_trips = trips.groupby("vehicle_number")
+    for vehicle_num, group in grouped_trips:
+        trips.loc[group.index, "max_charging"] = trips.loc[
+            group.index, "dwell_charging"
+        ].sum()
+        trips.loc[group.index, "required_charging"] = (
+            trips.loc[group.index, "trip_miles"].sum() * kwhmi
+        )
```
Not sure whether following is equivalent:
```python
trips["max_charging"] = trips["dwell_charging"].groupby("vehicle_number").transform('sum')
trips["required_charging"] = trips["trip_miles"].groupby("vehicle_number").transform('sum')
```

danlivengood replied:
The `kwhmi` multiplier would need to be included (suggested edit below, but @BainanXia, please confirm that this would work).  As far as their equivalence, it looks reasonable to me, but it's also taking Python code to a level I'm still learning.
```
trips["max_charging"] = trips["dwell_charging"].groupby("vehicle_number").transform('sum')
trips["required_charging"] = trips["trip_miles"].groupby("vehicle_number").transform('sum')*kwhmi
```

BainanXia replied:
Good catch. I think it should work with the multiplier as well.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/immediate_charging_HDV.py:
```diff
@@ -256,6 +256,25 @@ def immediate_charging(
+        trips["charging_allowed"] * trips["dwell_time"] * power * charging_efficiency
+    )
+
+    grouped_trips = trips.groupby("vehicle_number")
+    for vehicle_num, group in grouped_trips:
+        trips.loc[group.index, "max_charging"] = trips.loc[
+            group.index, "dwell_charging"
+        ].sum()
+        trips.loc[group.index, "required_charging"] = (
+            trips.loc[group.index, "trip_miles"].sum() * kwhmi
+        )
```
Same comment as above.

danlivengood replied:
Same reply from me from `immediate.py`

BainanXia commented on prereise/gather/demanddata/transportation_electrification/immediate.py:
```diff
@@ -217,7 +262,15 @@ def immediate_charging(
+    ]
+    print(np.sum(daily_resampled_profiles["weekend"]))
+    print(np.sum(daily_resampled_profiles["weekday"]))
+    print(output_load_sum_list)
```
We have a lot of prints throughout the refactor and some of them seems to be big chunks. I understand we would like to keep things around for warning/logging/debugging purpose but please still clean them up a little bit with corresponding message prompts, such as "warning", "sample output", etc. along with the prints to give users some info upon running.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -44,30 +49,37 @@ def smart_charging(
     """
 
+    # for code developer debugging printouts (1 = yes, 0 = no)
+    debug_printout = 0
```
Should we put this as another `kwarg` of the function then?

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -96,18 +108,12 @@ def smart_charging(
-        data_day = input_day
-        daily_vmt_total = data_helper.get_total_hdv_daily_vmt(newdata, veh_range)
+
+    external_signal = -min(external_signal) + external_signal
```
`external_signal -= min(external_signal)`

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -117,6 +123,90 @@ def smart_charging(
+            newdata["location_allowed"] = newdata["dwell_location"].isin(allowed)
+
+    # Add booleans for whether the trip_number (compared to total trips) allows charging
+    if trip_strategy == 1:
+        newdata["trip_allowed"] = True
+    elif trip_strategy == 2:
+        newdata["trip_allowed"] = newdata["trip_number"] == newdata["total_trips"]
```
Maybe we can raise a `ValueError` if the `trip_strategy` is not 1 or 2.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -117,6 +123,90 @@ def smart_charging(
+        * charging_efficiency
+    )
+
+    grouped_trips = newdata.groupby("vehicle_number")
+    for vehicle_num, group in grouped_trips:
+        newdata.loc[group.index, "max_charging"] = newdata.loc[
+            group.index, "dwell_charging"
+        ].sum()
+        newdata.loc[group.index, "required_charging"] = (
+            newdata.loc[group.index, "trip_miles"].sum() * kwhmi
+        )
```
Similar refactor suggestions as above.

danlivengood replied:
Same reply from me from `immediate.py`
In an ideal world, these filtering steps would be included in a separate function before `immediate` and `smart` charging are called.  That's a bonus refactor for later if there's time.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -117,6 +123,90 @@ def smart_charging(
+                    [filtered_census_data, individual], ignore_index=True
+                )
+            i += total_trips
+        del newdata
+        gc.collect()
```
Just curious, are we on a memory sensitive situation so that we would like to release the memory as soon as possible? Also we didn't call the garbage collector in the same logic during immediate charging above. Why in that case we simply want to remove the reference rather than reclaim the memory?

dmuldrew replied:
I think we can take this out, I was concerned that an error I was seeing earlier might be memory-related

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -136,8 +226,13 @@ def smart_charging(
+
     day_num = len(input_day)
     for day_iter in range(day_num):
+        print(f"Day: {day_iter}")
```
Should this output controlled by `debug_printout` as well, or we simply wanna track the progress of the loop, in which case `tqdm` can be used as elsewhere in the code base.

danlivengood replied:
This we would like to keep as a progress tracker for someone running the code.

BainanXia replied:
> This we would like to keep as a progress tracker for someone running the code.

Then we could use a `tqdm` to generate a progress bar instead of printing things out iteratively like [here](https://github.com/Breakthrough-Energy/PreREISE/blob/53fed977feced25bda68b49a4a4a6aa8e5c90626/prereise/gather/demanddata/eia/map_ba.py#L90)

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -166,6 +287,15 @@ def smart_charging(
                 # copy one vehicle information to the block
                 individual = newdata.iloc[i : i + total_trips].copy()
 
+                individual_vmt += float(
+                    individual.iloc[
+                        0, individual.columns.get_loc("total vehicle miles traveled")
+                    ]
+                )
```
Maybe it is equivalent to do
```python
individual_vmt += float(individual["total vehicle miles traveled"].iloc[0])
```

danlivengood replied:
This was something I added I believe, so if this is equivalent, I'm open to the change.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/smart_charging.py:
```diff
@@ -166,6 +287,15 @@ def smart_charging(
+                    ]
+                )
+                individual_trip_miles += float(
+                    individual.iloc[:, individual.columns.get_loc("trip_miles")].sum()
```
Maybe it is equivalent to do
```python
individual["trip_miles"].sum()
```

danlivengood replied:
Same: This was something I added I believe, so if this is equivalent, I'm open to the change.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/tests/test_immediate_charging_integration.py:
```diff
@@ -14,8 +15,118 @@
+    )
+
+
+def test_immediate_charging_hdv_1day():
```
Maybe we can merge this test with the previous one simply by plugging in different `veh_type` and `kwhmi`?

BainanXia commented on prereise/gather/demanddata/transportation_electrification/tests/test_immediate_charging_integration.py:
```diff
@@ -42,7 +153,7 @@ def test_immediate_charging_region1():
     daily_values = generate_daily_weighting(2017)
 
-    final_result = adjust_bev(
+    adjust_bev(
```
Now we just run it without checking the results?

BainanXia replied:
Same question below. If we keep these tests for integration test purpose given "unit" tests are added above, should we mark it as integration test with `@pytest.mark.integration`?

BainanXia commented on prereise/gather/demanddata/transportation_electrification/tests/test_immediate_charging_integration.py:
```diff
@@ -14,8 +15,118 @@
 )
 
 
+def test_immediate_charging_ldv_2days():
```
Just curious, why we do 2-day checking for `ldv` here, whereas `mdv` and `hdv` 1-day?

dmuldrew replied:
For LDV we need to check weekend and weekday values, while for MDV and HDV all days are the same and there is no such distinction

BainanXia replied:
I see. That makes sense.

BainanXia commented on prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging_integration.py:
```diff
@@ -5,15 +5,15 @@
 
 
-def test_smart_charging():
+def test_smart_charging_ldv_2days():
```
We might be able to merge the three tests by looping through several parameters but not sure that's worth the effort.

BainanXia reviewed (approved):
As discussed with @dmuldrew and @danlivengood during the meeting, this is good to go at this point after the linter is happy and the commit history is cleaned up. Thanks.
